flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/PulsarSourceBuilder.java

```diff
@@ -0,0 +1,615 @@
+                                            name,
+                                            configuration.getString(rawOption),
+                                            value);
+                                }
+                                configuration.setString(name, Objects.toString(value));
+                            }
+                        });
+        return this;
+    }
+
+    /**
+     * Build the {@link PulsarSource}.
+     *
+     * @return a PulsarSource with the settings made for this builder.
+     */
+    public PulsarSource<OUT> build() {
+        // Check builder configuration.
+        checkConfigurations(configuration);
+
+        SubscriptionType subscriptionType = configuration.get(PULSAR_SUBSCRIPTION_TYPE);
+
+        // Ensure the topic subscriber for pulsar.
+        if (subscriber == null) {
+            if (configuration.contains(PULSAR_TOPIC_NAMES)) {
+                List<String> topics =
+                        getOptionValue(
+                                configuration,
+                                PULSAR_TOPIC_NAMES,
+                                names -> PulsarJsonUtils.toList(String.class, names));
+                this.subscriber = PulsarSubscriber.getTopicListSubscriber(topics);
+            } else if (configuration.contains(PULSAR_TOPICS_PATTERN)) {
+                String topicPatternStr = configuration.get(PULSAR_TOPICS_PATTERN);
+                Pattern topicPattern = Pattern.compile(topicPatternStr);
+                RegexSubscriptionMode regexSubscriptionMode =
+                        configuration.get(PULSAR_REGEX_SUBSCRIPTION_MODE);
+
+                this.subscriber =
+                        PulsarSubscriber.getTopicPatternSubscriber(
+                                topicPattern, regexSubscriptionMode);
+            } else {
+                throw new IllegalArgumentException(
+                        "No "
+                                + PULSAR_TOPIC_NAMES.key()
+                                + " or "
+                                + PULSAR_TOPICS_PATTERN.key()
+                                + " was provided.");
+            }
+        }
+        if (subscriptionType == SubscriptionType.Key_Shared) {
+            if (rangeGenerator == null) {
+                LOG.warn(
+                        "No range generator provided for key_shared subscription,"
+                                + " we would use the DivideRangeGenerator as the default range generator.");
+                this.rangeGenerator = new UniformRangeGenerator();
+            }
+        } else {
+            // Override the range generator.
+            this.rangeGenerator = new FullRangeGenerator();
+        }
+
+        if (startCursorSupplier == null) {
+            this.startCursorSupplier = StartCursor::defaultStartCursor;
+        }
+
+        if (stopCursorSupplier == null) {
+            this.stopCursorSupplier = StopCursor::defaultStopCursor;
+        }
```

Comment:
I'd put these defaults in the field definition `private SerializableSupplier<StartCursor> startCursorSupplier = StartCursor::defaultStartCursor;`
Then we could also add `checkNotNull` to `setStartCursor`.
Finally, we could remove this validation.

Same for `StopCursor`.